From 739123196cc4bdb6eb1005d561cdf5987cc6d4f7 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Marek=20Beh=C3=BAn?= <marek.behun@nic.cz>
Date: Thu, 30 Apr 2020 00:18:17 +0200
Subject: [PATCH] mvebu: initial support for MOX on 5.4 kernel

This patch adds support for CZ.NIC Turris MOX (the modular router).

Specifications of module A (the base one):

- SoC: Marvell Armada 3720, ARMv8, dual-core 1 GHz
- RAM: 512 MB or 1 GB DDR3
- Flash: 8MB SPI NOR (w25q64dw)
- SDIO bus for Wi-Fi addon
- 1x USB 3.0 port
- 1x 1Gbps RJ45  port
- 1x RTC clock (Microchip MPC7940x)
- microSD card slot
- 34-pin GPIO CONNECTOR
- 8-pin PoE bracker
- 65-pin MOXTET to connect another modules

Other modules:
module B - miniPCIe slot
module C - 4x Gigabit Ethernet ports
module D - SFP module
module E - 8x Gigabit Ethernet ports
module G - miniPCIe slot
module F - 4x USB 3.0 ports
---
 package/boot/uboot-tools/files/mvebu          |   3 +
 .../base-files/etc/board.d/02_network         |  12 ++
 target/linux/mvebu/cortexa53/config-5.15       |  13 ++
 target/linux/mvebu/image/cortexa53.mk         |  12 ++
 ...is-mox-rwtm-support-ECDSA-signatures.patch | 158 ++++++++++++++++++
 ...Zero-max-link-speed-value-is-invalid.patch |  32 ++++
 ...x-implement-phylink-AN-restart-on-Se.patch | 137 +++++++++++++++
 7 files changed, 367 insertions(+)

diff --git a/package/boot/uboot-tools/files/mvebu b/package/boot/uboot-tools/files/mvebu
index 92b144330f..c298dfc63a 100644
--- a/package/boot/uboot-tools/files/mvebu
+++ b/package/boot/uboot-tools/files/mvebu
@@ -24,6 +24,9 @@ cznic,turris-omnia)
 		ubootenv_add_uci_config "/dev/mtd0" "0xf0000" "0x10000" "0x10000"
 	fi
 	;;
+cznic,turris-mox)
+	ubootenv_add_uci_config "/dev/mtd2" "0x0" "0x00010000"
+	;;
 glinet,gl-mv1000)
 	ubootenv_add_uci_config "/dev/mtd1" "0x0" "0x8000" "0x8000" "1"
 	;;
diff --git a/target/linux/mvebu/cortexa53/base-files/etc/board.d/02_network b/target/linux/mvebu/cortexa53/base-files/etc/board.d/02_network
index c945251e4e..ad76a31ba9 100755
--- a/target/linux/mvebu/cortexa53/base-files/etc/board.d/02_network
+++ b/target/linux/mvebu/cortexa53/base-files/etc/board.d/02_network
@@ -11,6 +11,18 @@ board_config_update
 board=$(board_name)
 
 case "$board" in
+cznic,turris-mox)
+	mox_lan_interfaces=""
+	for net in /sys/class/net/lan*; do
+		[ -e "$net" ] || continue
+		append mox_lan_interfaces "${net##*/}"
+	done
+	if [ -z "$mox_lan_interfaces" ]; then
+		ucidef_set_interface_lan "eth0"
+	else
+		ucidef_set_interfaces_lan_wan "$mox_lan_interfaces" "eth0"
+	fi
+	;;
 glinet,gl-mv1000|\
 globalscale,espressobin|\
 globalscale,espressobin-emmc|\
diff --git a/target/linux/mvebu/cortexa53/config-5.15 b/target/linux/mvebu/cortexa53/config-5.15
index 3c195291d1..027b2a9b34 100644
--- a/target/linux/mvebu/cortexa53/config-5.15
+++ b/target/linux/mvebu/cortexa53/config-5.15
@@ -45,6 +45,10 @@ CONFIG_GENERIC_PINCONF=y
 CONFIG_ILLEGAL_POINTER_VALUE=0xdead000000000000
 CONFIG_MAILBOX=y
 # CONFIG_MAILBOX_TEST is not set
++CONFIG_I2C_PXA=y
++# CONFIG_I2C_PXA_SLAVE is not set
+CONFIG_GPIO_MOXTET=y
+CONFIG_MOXTET=y
 CONFIG_MFD_SYSCON=y
 CONFIG_MMC_SDHCI_XENON=y
 CONFIG_MODULES_USE_ELF_RELA=y
index 6a3a568655..ab694ab4fb 100644
--- a/target/linux/mvebu/image/cortexa53.mk
+++ b/target/linux/mvebu/image/cortexa53.mk
@@ -1,3 +1,13 @@
+define Device/cznic_turris-mox
+  $(call Device/Default-arm64)
+  DEVICE_VENDOR := CZ.NIC
+  DEVICE_MODEL := Turris MOX
+  DEVICE_PACKAGES := kmod-usb2 kmod-gpio-button-hotplug kmod-rtc-ds1307 kmod-i2c-pxa
+  SOC := armada-3720
+  BOOT_SCRIPT := turris-mox
+endef
+TARGET_DEVICES += cznic_turris-mox
+
 define Device/glinet_gl-mv1000
   $(call Device/Default-arm64)
   DEVICE_VENDOR := GL.iNet
-- 
2.34.1

